Initial commit to fetch monthly active UTXOs

eswarasai opened the pull request:
### Description
- [x] Added `serverless` setup with full support of local development
- [x] Added `get` endpoint and `update` cron to update the stats on a daily basis
- [x] Update README.md with setup instructions

Fixes: https://github.com/leapdao/meta/issues/61
Fixes: #2 

### TODO:
- [x] Added plasma provider as environment variable
- [x] Integrated the API endpoint on the website

eswarasai committed:
Initial commit to fetch monthly active UTXOs

MaxStalker commented on src/update.ts:
```diff
@@ -0,0 +1,69 @@
+  const unspents: import('./types').UnspentWithTx[] = await Promise.all(
+    uTxos.map((u: import('./types').UnspentWithTx) =>
+      web3Plasma.eth
+        .getTransaction(bufferToHex(u.outpoint.hash))
```
would this block work with `await` here instead of chaining? To keep the consistency in style? :)

eswarasai replied:
Good idea. I'll update this. Thanks for pointing out 👍

MaxStalker commented on src/update.ts:
```diff
@@ -0,0 +1,69 @@
+          return u;
+        })
+        .then((uTx: import('./types').UnspentWithTx) =>
+          web3Plasma.eth
```
We can cache blocks that we already requested with simple object, stored outside of this mapping. This will speed up the process and lower number of requests to node. Potentially the bill for using this function as well :)

eswarasai replied:
Not sure I get the idea of what you're suggesting here. You'd like me to save the unspent transactions in a DB table and just keep updating the new one's to process the stats?

I think my actual question is, wouldn't the old unspent transactions ever become spent/exit? I'm still trying to get my head around this 😓

MaxStalker replied:
I was refering to block fetching. You are building an array of promises, and then waiting until all of the resolve, right? I suppose there can be cases, when we have several transactions in the same block, thus no need to refetch the block is we already done this before, right?

eswarasai replied:
Thanks for the pointer. I'll spend a bit more time thinking about this to see a way to optimize it right from the start 😅

eswarasai commented on serverless.yml:
```diff
@@ -0,0 +1,78 @@
+service:
+  name: leap-stats
+  app: leap-stats
+  org: eswarasai
```
This should be of LeapDAO, but I don't have access to the serverless account. So temporarily using mine 😅

eswarasai committed:
Refactored fetching timestamps for blocks

eswarasai commented:
@MaxStalker - Just updated with the suggested change of fetching block numbers timestamps. Let me know your final review. Thanks!

MaxStalker commented on README.md:
````diff
@@ -1 +1,34 @@
+```
+
+### Deployment
+- Configure AWS Lambda account using [`aws-cli`](https://serverless.com/framework/docs/providers/aws/guide/credentials#setup-with-the-aws-cli) as per the instructions in the link
````
Probably will be better to put link under the word link, instead of aws-cli - me, for example, had hard time to noticing where it is ^__^

eswarasai replied:
Sure, I'll update that 😅

eswarasai commented on src/update.ts:
```diff
@@ -0,0 +1,72 @@
+
+  // Filter current month UTXOs
+  const endTimestamp = new Date().getTime() / 1000;
+  const startTimestamp = endTimestamp - 30 * 24 * 60 * 60; // 30 days from now
```
Lol, yeah. Initially I did it for the previous month. But later just decided to shift it for last 30 days. I'll need to update the comments, missed that part 🤦‍♂

eswarasai commented on src/update.ts:
```diff
@@ -0,0 +1,72 @@
+
+  const params = {
+    TableName: process.env.STATS_TABLE,
+    Item: {
```
Since we're not fetching the stats of only previous month, it's that of last 30 days. So adding a month field might not make sense here.

eswarasai commented on serverless.yml:
```diff
@@ -0,0 +1,78 @@
+    handler: src/update.updateStats
+    events:
+      - schedule:
+          rate: cron(0 0 * * ? *)
```
@MaxStalker - Just wanted to point out here that the update stats runs as a scheduler every 24 hours 🙂

eswarasai committed:
Updated comments

troggy commented:
@MaxStalker @eswarasai guys, why it is not merged yet? Still work in progress?

MaxStalker commented:
@troggy I believe there are 2 more todo items that @eswarasai wanted to finish.

troggy commented:
I don't see them a part of bounty description

MaxStalker commented:
@troggy `Make plasma provider configurable based on environment` can be a part of `Deliverables` for deploying it for testnet and mainnet? :)

eswarasai committed:
Added PLASMA_PROVIDER env variable

eswarasai committed:
Enabled CORS for GET API

eswarasai commented:
@MaxStalker - I've updated things here as well as the README doc for the deployment. There were some issues with the cron not running for the update stats function. Hence I've added logs for debugging. After deploying things on the LeapDAO AWS environment and we ensure that everything is working fine, we can remove the logs and re-deploy. Let me know whenever you're around to deploy and test things :)

troggy commented on src/update.ts:
```diff
@@ -0,0 +1,80 @@
+  const currentUtxos = unspents.filter(u => u.timestamp >= startTimestamp && u.timestamp <= endTimestamp);
+
+  // Remove same address transactions
+  const activeUtxos = currentUtxos.filter(c => c.transaction.from !== c.transaction.to);
```
Keep in mind, this works only for "simple" transfer txs. It considers that destination of tx is the very first output, which is not always correct — transaction may split the input between different beneficiaries, spending condition is tricky etc

Similar logic applies to `from` field (see same links)

Reference:
[1] - How we set `to` for transaction we return from JSON RPC in leap-node https://github.com/leapdao/leap-node/blob/23f807ada23c53eb5d7cb2d371e960ef8995a3fe/src/api/methods/txResponse.js#L23
[2] - the code which actually derives `to` attribute: https://github.com/leapdao/leap-core/blob/eec71e1451100fcf2ddaa8f5a7a5722d7fb6bd41/lib/transaction.js#L434

troggy replied:
just a note. I don't know how to reliably solve this, so it is the best we can have (easily)

troggy committed:
we use eu-west-1 region

troggy commented:
Deployment: https://94m4ns21fi.execute-api.eu-west-1.amazonaws.com/dev/stats/get

troggy committed:
use fixed leap-core and getUnspentAll. getUnspent() didn't work correct

troggy commented:
Deployed here: https://zy2mo76rd6.execute-api.eu-west-1.amazonaws.com/mainnet/stats/get

Update seems to work:

```
2019-11-26 17:21:06.188 (+03:00)	7c86cf4f-f35d-44c5-afba-b535e03f7824	INFO	Updating stats... wss://mainnet-node1.leapdao.org:1443
2019-11-26 17:21:06.511 (+03:00)	7c86cf4f-f35d-44c5-afba-b535e03f7824	INFO	Active UTXOs metric 6
2019-11-26 17:21:06.512 (+03:00)	7c86cf4f-f35d-44c5-afba-b535e03f7824	INFO	Saving to DynamoDB...
END RequestId: 7c86cf4f-f35d-44c5-afba-b535e03f7824
```

but not saving to DynamoDb anything: no records in db, `/get` endpoint returns 0 as well.

am I doing something wrong? How did you test it guys @MaxStalker @eswarasai ?

troggy committed:
add helper script for mainnet deployment

troggy commented:
I see.. the problem may be with using `callback` and `async` at the same time. Will try to fix

troggy committed:
speed up by not awaiting in the loop/map

troggy committed:
fix: don't use callback and async at the same time — callback will be ignored

troggy commented:
Fixed here: https://github.com/leapdao/leap-stats/pull/1/commits/be519aa03db2dc18c0542b89d69ed5ee8505c0b1

troggy committed:
chore: optimize code a bit. Replacing 3 loops with 1

troggy commented:
@eswarasai I suggest the following optimization to the code, what do you think? https://github.com/leapdao/leap-stats/pull/3

The idea is to squash three loops (1 map, 2 filters) into just 1 loop (filter)

Totally premature 😅

eswarasai committed:
Merge pull request #3 from leapdao/chore/simplify-code

chore: optimize code a bit. Replacing 3 loops with 1

eswarasai commented:
@troggy @MaxKuck - I think it's good to be merged now. What do you guys think? Anything else pending? Thanks!

troggy commented:
great work @eswarasai 🥦
